From 2035bfba3d67c2d939bfc91da517ac76bb7c4677 Mon Sep 17 00:00:00 2001
From: Chen Baozi <chenbaozi@phytium.com.cn>
Date: Mon, 17 Jun 2024 19:33:00 +0800
Subject: [PATCH 010/150] dt-bindings: gpio: Add bindings for Phytium GPIO

This patch documents the DT bindings for Phytium GPIO controller.

Signed-off-by: Chen Baozi <chenbaozi@phytium.com.cn>
Change-Id: I0862f0393af1380b9f7bace1e08fd46c7cfc326f
Signed-off-by: Andrew Powers-Holmes <aholmes@omnom.net>
---
 .../bindings/gpio/phytium,gpio.yaml           | 114 ++++++++++++++++++
 MAINTAINERS                                   |   1 +
 2 files changed, 115 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/gpio/phytium,gpio.yaml

diff --git a/Documentation/devicetree/bindings/gpio/phytium,gpio.yaml b/Documentation/devicetree/bindings/gpio/phytium,gpio.yaml
new file mode 100644
index 000000000000..4a2d586d1880
--- /dev/null
+++ b/Documentation/devicetree/bindings/gpio/phytium,gpio.yaml
@@ -0,0 +1,114 @@
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/gpio/phytium,gpio.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Phytium GPIO controller
+
+description: |
+  Phytium GPIO controllers have one or two configurable ports, each of which
+  are intended to be represented as child nodes with the generic GPIO-controller
+  properties as desribed in this bindings file.
+
+maintainers:
+  - Chen Baozi <chenbaozi@phytium.com.cn>
+
+properties:
+  $nodename:
+    pattern: "^gpio@[0-9a-f]+$"
+
+  compatible:
+    const: phytium,gpio
+
+  reg:
+    maxItems: 1
+
+  gpio-controller: true
+
+  "#address-cells":
+    const: 1
+
+  "#size-cells":
+    const: 0
+
+  '#gpio-cells':
+    const: 2
+
+  interrupts:
+    description: |
+      The interrupts to the parent controller raised when GPIOs generate
+      the interrupts. If the controller provides one combined interrupt
+      for all GPIOs, specify a single interrupt. If the controller provides
+      one interrupt for each GPIO, provide a list of interrupts that
+      correspond to each of the GPIO pins.
+    minItems: 1
+    maxItems: 32
+
+  interrupt-controller: true
+
+  '#interrupt-cells':
+    const: 2
+
+patternProperties:
+  "^gpio-(port|controller)@[0-9a-f]+$":
+    type: object
+    properties:
+      compatible:
+        const: phytium,gpio-port
+
+      reg:
+        maxItems: 1
+
+      nr-gpios:
+        $ref: /schemas/types.yaml#/definitions/uint32
+        description: The number of GPIO pins exported by the port.
+        default: 32
+        minimum: 1
+        maximum: 32
+
+    required:
+      - compatible
+      - reg
+      - gpio-controller
+      - '#gpio-cells'
+
+    dependencies:
+      interrupt-controller: [ interrupts ]
+
+    additionalProperties: false
+
+additionalProperties: false
+
+required:
+  - compatible
+  - reg
+  - "#address-cells"
+  - "#size-cells"
+
+examples:
+  - |
+    gpio: gpio@28004000 {
+      compatible = "phytium,gpio";
+      reg = <0x0 0x28004000 0x0 0x1000>;
+      gpio-controller;
+      #gpio-cells = <2>;
+      #address-cells = <1>;
+      #size-cells = <0>;
+      interrupts = <GIC_SPI 10 IRQ_TYPE_LEVEL_HIGH>;
+      interrupt-controller;
+      #interrupt-cells = <2>;
+
+      porta: gpio-port@0 {
+        compatible = "phytium,gpio-port";
+        reg = <0>;
+        nr-gpios = <8>;
+      };
+
+      portb: gpio-port@1 {
+        compatible = "phytium,gpio-port";
+        reg = <1>;
+        nr-gpios = <8>;
+      };
+};
+...
diff --git a/MAINTAINERS b/MAINTAINERS
index 7f1a1fd31090..37d5a89c026a 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -2530,6 +2530,7 @@ ARM/PHYTIUM SOC SUPPORT
 M:	Chen Baozi <chenbaozi@phytium.com.cn>
 S:	Maintained
 W:	https://www.phytium.com.cn
+F:	Documentation/devicetree/bindings/gpio/phytium,gpio.yaml
 F:	arch/arm64/boot/dts/phytium/*
 
 ARM/QUALCOMM CHROMEBOOK SUPPORT
-- 
2.47.0

